docs/how_to_guides/configuring.rst
.. _how_to_configure:

Configuring
===========

Hypercorn is configured via a command line arguments, or via a
:class:`hypercorn.config.Config` instance, which can be created
manually, loaded from a TOML, Python file, Python module, or a
dictionary instance.

Via a TOML file
---------------

`TOML <https://github.com/toml-lang/toml>`_ is the prefered format for
Hypercorn configuration files. Files in this format can be loaded via
the command line using the ``-c``, ``--config`` option e.g.

.. code-block::

     hypercorn --config file_path/file_name.toml

To load programatically :meth:`hypercorn.config.Config.from_toml` can
be used,

.. code-block:: python

     config = Config()
     config.from_toml("file_path/file_name.toml")

Via a Python module
-------------------

A Python module or an instance within a python module can be used to
configure Hypercorn. In both cases the attributes matching
configuration values will be used. This can be specified via the
command line using the ``-c``, ``--config`` option with the
``python:`` prefix e.g.

.. code-block::

     hypercorn --config python:module_name

To load programatically :meth:`hypercorn.config.Config.from_object`
can be used,

.. code-block:: python

     config = Config()
     config.from_object("module_name.instance")

Via a Python file
-------------------

A Python file can be loaded and the attributes matching configuration
values used to configure Hypercorn. This can be specified via the
command line using the ``-c``, ``--config`` option with the
``file:`` prefix e.g.

.. code-block::

     hypercorn --config file:file_path/file_name.py

To load programatically :meth:`hypercorn.config.Config.from_pyfile`
can be used,

.. code-block:: python

     config = Config()
     config.from_pyfile("file_path/file_name.py")

Configuration options
=====================

========================== ============================= =============================================== ========================
Attribute                  Command line                  Purpose                                         Default
-------------------------- ----------------------------- ----------------------------------------------- ------------------------
access_log_format          ``--access-logformat``        The log format for the access log, see
                                                         :ref:`how_to_log`.
accesslog                  ``--access-logfile``          The target logger for access logs, use
                                                         ``-`` for stdout.
alpn_protocols             N/A                           The HTTP protocols to advertise over            ``h2`` and ``http/1.1``
                                                         ALPN.
alt_svc_headers            N/A                           List of header values to return as
                                                         Alt-Svc headers.
application_path           N/A                           The path location of the ASGI                   cwd
                                                         application.
backlog                    ``--backlog``                 The maximum number of pending                   100
                                                         connections.
bind                       ``-b``, ``--bind``            The TCP host/address to bind to.
                                                         Should be either host:port, host,
                                                         unix:path or fd://num, e.g.
                                                         127.0.0.1:5000, 127.0.0.1,
                                                         unix:/tmp/socket or fd://33
                                                         respectively.
ca_certs                   ``--ca-certs``                Path to the SSL CA certificate file.
certfile                   ``--certfile``                Path to the SSL certificate file.
ciphers                    ``--ciphers``                 Ciphers to use for the SSL setup.               ``ECDHE+AESGCM``
debug                      ``--debug``                   Enable debug mode, i.e. extra logging           ``False``
                                                         and checks.
dogstatsd_tags             N/A                           DogStatsd format tag, see
                                                         :ref:`using_statsd`.
errorlog                   ``--error-logfile``           The target location for the error log,
                           ``--log-file``                use `-` for stderr.
graceful_timeout           ``--graceful-timeout``        Time to wait after SIGTERM or Ctrl-C            3s
                                                         for any remaining requests (tasks) to
read_timeout               ``--read-timeout``            Seconds to wait before timing out reads         No timeout.
                                                         on TCP sockets.
group                      ``-g``, ``--group``           Group to own any unix sockets.
h11_max_incomplete_size    N/A                           The max HTTP/1.1 request line + headers         16KiB
                                                         size in bytes.
h11_pass_raw_headers       N/A                           Pass the raw headers from h11 to the            ``False``
                                                         Request object, which preserves header
                                                         casing.
h2_max_concurrent_streams  N/A                           Maximum number of HTTP/2 concurrent             100
                                                         streams.
h2_max_header_list_size    N/A                           Maximum number of HTTP/2 headers.               65536
h2_max_inbound_frame_size  N/A                           Maximum size of a HTTP/2 frame.                 16KiB
include_date_header        N/A                           Include the                                     ``True``
                                                         ``Date: Tue, 15 Nov 1994 08:12:31 GMT``
                                                         header.
include_server_header      N/A                           Include the ``Server: Hypercorn`` header.       ``True``
insecure_bind              ``--insecure-bind``           The TCP host/address to bind to. SSL
                                                         options will not apply to these binds.
                                                         See *bind* for formatting options.
                                                         Care must be taken! See HTTP -> HTTPS
                                                         redirection docs.
keep_alive_max_requests    N/A                           Maximum number of requests before connection    1000
                                                         is closed. HTTP/1 & HTTP/2 only.
keep_alive_timeout         ``--keep-alive``              Seconds to keep inactive connections alive      5s
                                                         before closing.
keyfile                    ``--keyfile``                 Path to the SSL key file.
keyfile_password           ``--keyfile-password``        Password for the keyfile if the keyfile is
                                                         password-protected.
logconfig                  ``--log-config``              A Python logging configuration file. This       The logging ini format.
                                                         can be prefixed with 'json:' or 'toml:' to
                                                         load the configuration from a file in that
                                                         format.
logconfig_dict             N/A                           A Python logging configuration dictionary.
logger_class               N/A                           Type of class to use for logging.
loglevel                   ``--log-level``               The (error) log level.                          ``INFO``
max_app_queue_size         N/A                           The maximum number of events to queue up        10
                                                         sending to the ASGI application.
max_requests               ``--max-requests``            Maximum number of requests a worker will
                                                         process before restarting.
max_requests_jitter        ``--max-requests-jitter``     This jitter causes the max-requests per worker  0
                                                         to be randomized by
                                                         ``randint(0, max_requests_jitter)``
pid_path                   ``-p``, ``--pid``             Location to write the PID (Program ID) to.
quic_bind                  ``--quic-bind``               The UDP/QUIC host/address to bind to. See
                                                         *bind* for formatting options.
root_path                  ``--root-path``               The setting for the ASGI root_path
                                                         variable.
server_names               ``--server-name``             The hostnames that can be served, requests
                                                         to different hosts will be responded to
                                                         with 404s.
shutdown_timeout           N/A                           Timeout when waiting for Lifespan               60s
                                                         shutdowns to complete.
ssl_handshake_timeout      N/A                           Timeout when waiting for SSL handshakes to      60s
                                                         complete.
startup_timeout            N/A                           Timeout when waiting for Lifespan               60s
                                                         startups to complete.
statsd_host                ``--statsd-host``             The host:port of the statsd server.
statsd_prefix              ``--statsd-prefix``           Prefix for all statsd messages.
umask                      ``-m``, ``--umask``           The permissions bit mask to use on any
                                                         unix sockets.
use_reloader               ``--reload``                  Enable automatic reloads on code changes.
user                       ``-u``, ``--user``            User to own any unix sockets.
verify_flags               N/A                           SSL context verify flags.
verify_mode                ``--verify-mode``             SSL verify mode for peer's certificate,
                                                         see ssl.VerifyMode enum for possible
                                                         values.
websocket_max_message_size N/A                           Maximum size of a WebSocket frame.              16MiB
websocket_ping_interval    ``--websocket-ping-interval`` If set this is the time in seconds between
                                                         pings sent to the client. This can be used
                                                         to keep the websocket connection alive.
worker_class               ``-k``, ``--worker-class``    The type of worker to use. Options include
                                                         asyncio, uvloop (pip install
                                                         hypercorn[uvloop]), and trio (pip install
                                                         hypercorn[trio]).
workers                    ``-w``, ``--workers``         The number of workers to spawn and use.         1
wsgi_max_body_size         N/A                           The maximum size of a body that will be         16MiB
                                                         accepted in WSGI mode.
========================== ============================= =============================================== ========================


CHANGELOG.rst
0.17.3 2024-05-28
-----------------

* Restore set TCP_NODELAY on TCP sockets
* Support uvloop >= 0.18 and the loop_factory argument
* Bugfix ensure ExceptionGroup lifespan failures crash the server.

0.17.2 2024-05-27
-----------------

* Bugfix pass the correct quic connection to the H3 Protocol.

0.17.1 2024-05-27
-----------------

* Bugfix revert set TCP_NODELAY on sockets.

0.17.0 2024-05-27
-----------------

* Set TCP_NODELAY on sockets.
* Support sending trailing headers on h2/h3.
* Add support for lifespan state.
* Allow sending of the response before body data arrives.
* Bugfix properly set host header to ascii string in
  ProxyFixMiddleware.
* Bugfix encode headers using latin-1.
* Bugfix don't double-access log if the response was sent.
* Bugfix a statsd logging bug.
* Bugfix handle already-closed on StreamEnded.
* Bugfix send a 400 response if data is received before the websocket
  is accepted.
* Bugfix ensure only a single QUIC timer task per connection.
* Bugfix ensure responses are sent with empty bodies for WSGI.

0.16.0 2024-01-01
-----------------

* Add a max keep alive requests configuration option, this mitigates
  the HTTP/2 rapid reset attack.
* Return subprocess exit code if non-zero.
* Add ProxyFix middleware to make it easier to run Hypercorn behind a
  proxy.
* Support restarting workers after max requests to make it easier to
  manage memory leaks in apps.
* Bugfix ensure the idle task is stopped on error.
* Bugfix revert autoreload error because reausing old sockets.
* Bugfix send the hinted error from h11 on RemoteProtocolErrors.
* Bugfix handle asyncio.CancelledError when socket is closed without
  flushing.
* Bugfix improve WSGI compliance by closing iterators, only sending
  headers on first response byte, erroring if ``start_response`` is
  not called, and switching wsgi.errors to stdout.
* Don't error on LocalProtoclErrors for ws streams to better cope with
  race conditions.

0.15.0 2023-10-29
-----------------

* Improve the NoAppError to help diagnose why the app has not been
  found.
* Log cancelled requests as well as successful to aid diagnositics of
  failures.
* Use more modern asyncio apis. This will hopefully fix reported
  memory leak issues.
* Bugfix only load the application in the main process if the reloader
  is being used.
* Bugfix Autoreload error because reausing old sockets.
* Bugfix scope client usage for sock binding.
* Bugfix disable multiprocessing if number of workers is 0 to support
  systems that don't support multiprocessing.

0.14.4 2023-07-08
-----------------

* Bugfix Use tomllib/tomli for .toml support replacing the
  unmaintained toml library.
* Bugfix server hanging on startup failure.
* Bugfix close websocket with 1011 on internal error (1006 is a
  client-only code).
* Bugfix support trio > 0.22 utilising exception groups (note trio <=
  0.22 is not supported).
* Bugfix except ConnectionAbortedError which can be raised on Windows
  machines.
* Bugfix ensure that closed is sent on reading end.
* Bugfix handle read_timeout exception on trio.
* Support and test against Python 3.11.
* Add explanation of PicklingErrors.
* Add config option to pass raw h11 headers.

0.14.3 2022-09-04
-----------------

* Revert Preserve response headers casing for HTTP/1 as this breaks
  ASGI frameworks.
* Bugfix stream WSGI responses

0.14.2 2022-09-03
-----------------

* Bugfix add missing ASGI version to lifespan scope.
* Bugfix preserve the HTTP/1 request header casing through to the ASGI
  app.
* Bugifx ensure the config loglevel is respected.
* Bugfix ensure new processes are spawned not forked.
* Bugfix ignore dunder vars in config objects.
* Bugfix clarify the subprotocol exception.

0.14.1 2022-08-29
-----------------

* Fix Python3.7 compatibility.

0.14.0 2022-08-29
-----------------

* Bugfix only recycle a HTTP/1.1 connection if client is DONE.
* Bugfix uvloop may raise a RuntimeError.
* Bugfix ensure 100ms sleep between Windows workers starting.
* Bugfix ensure lifespan shutdowns occur.
* Bugfix close idle Keep-Alive connections on graceful exit.
* Bugfix don't suppress 412 bodies.
* Bugfix don't idle close upgrade requests.
* Allow control over date header addition.
* Allow for logging configuration to be loaded from JSON or TOML
  files.
* Preserve response headers casing for HTTP/1.
* Support the early hint ASGI-extension.
* Alter the process and reloading system such that it should work
  correctly in all configurations.
* Directly support serving WSGI applications (and drop support for
  ASGI-2, now ASGI-3 only).

0.13.2 2021-12-23
-----------------

* Bugfix re-enable HTTP/3.

0.13.1 2021-12-16
-----------------

* Bugfix trio tcp server read completion.

0.13.0 2021-12-14
-----------------

* Bugfix eof and keep alive handling.
* Bugfix Handle SSLErrors when reading.
* Support websocket close reasons.
* Improve the graceful shutdown, such that it works as expected.
* Support a keyfile password argument.
* Change the logging level to warning for lifespan not supported.
* Shutdown the default executor.
* Support additional headers for WS accept response.

0.12.0 2021-11-08
-----------------

* Correctly utilise SCRIPT_NAME in the wsgi middleware.
* Support Python 3.10.
* Support badly behaved HTTP/2 clients that omit a :authority header
  but provide a host header.
* Use environment marker for uvloop (on windows).
* Use StringIO and BytesIO for more performant websocket buffers.
* Add optional read timeout.
* Rename errors to add a ``Error`` suffix, most notably
  ``LifespanFailure`` to ``LifespanFailureError``.
* Bugfix ensure keep alive timeout is cancelled on closure.
* Bugfix statsd type error.
* Bugfix prevent spawning whilst a task group is exit(ing).

0.11.2 2021-01-10
-----------------

* Bugfix catch the base class ConnectionError.
* Bugfix catch KeyboardInterrupt if raised here e.g. on Windows.
* Bugfix support non-standard HTTP status codes in the access logger.

src/hypercorn/config.py
from __future__ import annotations

import importlib
import importlib.util
import logging
import os
import socket
import stat
import sys
import types
import warnings
from dataclasses import dataclass
from ssl import (
    create_default_context,
    OP_NO_COMPRESSION,
    Purpose,
    SSLContext,
    TLSVersion,
    VerifyFlags,
    VerifyMode,
)
from time import time
from typing import Any, AnyStr, Dict, List, Mapping, Optional, Tuple, Type, Union
from wsgiref.handlers import format_date_time

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

from .logging import Logger

BYTES = 1
OCTETS = 1
SECONDS = 1.0

FilePath = Union[AnyStr, os.PathLike]
SocketKind = Union[int, socket.SocketKind]


@dataclass
class Sockets:
    secure_sockets: List[socket.socket]
    insecure_sockets: List[socket.socket]
    quic_sockets: List[socket.socket]


class SocketTypeError(Exception):
    def __init__(self, expected: SocketKind, actual: SocketKind) -> None:
        super().__init__(
            f'Unexpected socket type, wanted "{socket.SocketKind(expected)}" got '
            f'"{socket.SocketKind(actual)}"'
        )


class Config:
    _bind = ["127.0.0.1:8000"]
    _insecure_bind: List[str] = []
    _quic_bind: List[str] = []
    _quic_addresses: List[Tuple] = []
    _log: Optional[Logger] = None
    _root_path: str = ""
    _daemon: bool = True

    access_log_format = '%(h)s %(l)s %(l)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"'
    accesslog: Union[logging.Logger, str, None] = None
    alpn_protocols = ["h2", "http/1.1"]
    alt_svc_headers: List[str] = []
    application_path: str
    backlog = 100
    ca_certs: Optional[str] = None
    certfile: Optional[str] = None
    ciphers: str = "ECDHE+AESGCM"
    debug = False
    dogstatsd_tags = ""
    errorlog: Union[logging.Logger, str, None] = "-"
    graceful_timeout: float = 3 * SECONDS
    read_timeout: Optional[int] = None
    group: Optional[int] = None
    h11_max_incomplete_size = 16 * 1024 * BYTES
    h11_pass_raw_headers = False
    h2_max_concurrent_streams = 100
    h2_max_header_list_size = 2**16
    h2_max_inbound_frame_size = 2**14 * OCTETS
    include_date_header = True
    include_server_header = True
    keep_alive_timeout = 5 * SECONDS
    keep_alive_max_requests = 1000
    keyfile: Optional[str] = None
    keyfile_password: Optional[str] = None
    logconfig: Optional[str] = None
    logconfig_dict: Optional[dict] = None
    logger_class = Logger
    loglevel: str = "INFO"
    max_app_queue_size: int = 10
    max_requests: Optional[int] = None
    max_requests_jitter: int = 0
    pid_path: Optional[str] = None
    server_names: List[str] = []
    shutdown_timeout = 60 * SECONDS
    ssl_handshake_timeout = 60 * SECONDS
    startup_timeout = 60 * SECONDS
    statsd_host: Optional[str] = None
    statsd_prefix = ""
    umask: Optional[int] = None
    use_reloader = False
    user: Optional[int] = None
    verify_flags: Optional[VerifyFlags] = None
    verify_mode: Optional[VerifyMode] = None
    websocket_max_message_size = 16 * 1024 * 1024 * BYTES
    websocket_ping_interval: Optional[float] = None
    worker_class = "asyncio"
    workers = 1
    wsgi_max_body_size = 16 * 1024 * 1024 * BYTES

    def set_cert_reqs(self, value: int) -> None:
        warnings.warn("Please use verify_mode instead", Warning)
        self.verify_mode = VerifyMode(value)

    cert_reqs = property(None, set_cert_reqs)

    @property
    def daemon(self) -> bool:
        return self._daemon

    @daemon.setter
    def daemon(self, value: bool) -> None:
        self._daemon = value

    @property
    def log(self) -> Logger:
        if self._log is None:
            self._log = self.logger_class(self)
        return self._log

    @property
    def bind(self) -> List[str]:
        return self._bind

    @bind.setter
    def bind(self, value: Union[List[str], str]) -> None:
        if isinstance(value, str):
            self._bind = [value]
        else:
            self._bind = value

    @property
    def insecure_bind(self) -> List[str]:
        return self._insecure_bind

    @insecure_bind.setter
    def insecure_bind(self, value: Union[List[str], str]) -> None:
        if isinstance(value, str):
            self._insecure_bind = [value]
        else:
            self._insecure_bind = value

    @property
    def quic_bind(self) -> List[str]:
        return self._quic_bind

    @quic_bind.setter
    def quic_bind(self, value: Union[List[str], str]) -> None:
        if isinstance(value, str):
            self._quic_bind = [value]
        else:
            self._quic_bind = value

    @property
    def root_path(self) -> str:
        return self._root_path

    @root_path.setter
    def root_path(self, value: str) -> None:
        self._root_path = value.rstrip("/")

    def create_ssl_context(self) -> Optional[SSLContext]:
        if not self.ssl_enabled:
            return None

        context = create_default_context(Purpose.CLIENT_AUTH)
        context.set_ciphers(self.ciphers)
        context.minimum_version = TLSVersion.TLSv1_2  # RFC 7540 Section 9.2: MUST be TLS >=1.2
        context.options = OP_NO_COMPRESSION  # RFC 7540 Section 9.2.1: MUST disable compression
        context.set_alpn_protocols(self.alpn_protocols)

        if self.certfile is not None and self.keyfile is not None:
            context.load_cert_chain(
                certfile=self.certfile,
                keyfile=self.keyfile,
                password=self.keyfile_password,
            )

        if self.ca_certs is not None:
            context.load_verify_locations(self.ca_certs)
        if self.verify_mode is not None:
            context.verify_mode = self.verify_mode
        if self.verify_flags is not None:
            context.verify_flags = self.verify_flags

        return context

    @property
    def ssl_enabled(self) -> bool:
        return self.certfile is not None and self.keyfile is not None

    def create_sockets(self) -> Sockets:
        if self.ssl_enabled:
            secure_sockets = self._create_sockets(self.bind)
            insecure_sockets = self._create_sockets(self.insecure_bind)
            quic_sockets = self._create_sockets(self.quic_bind, socket.SOCK_DGRAM)
            self._set_quic_addresses(quic_sockets)
        else:
            secure_sockets = []
            insecure_sockets = self._create_sockets(self.bind)
            quic_sockets = []
        return Sockets(secure_sockets, insecure_sockets, quic_sockets)

    def _set_quic_addresses(self, sockets: List[socket.socket]) -> None:
        self._quic_addresses = []
        for sock in sockets:
            name = sock.getsockname()
            if type(name) is not str and len(name) >= 2:
                self._quic_addresses.append(name)
            else:
                warnings.warn(
                    f'Cannot create a alt-svc header for the QUIC socket with address "{name}"',
                    Warning,
                )

    def _create_sockets(
        self, binds: List[str], type_: int = socket.SOCK_STREAM
    ) -> List[socket.socket]:
        sockets: List[socket.socket] = []
        for bind in binds:
            binding: Any = None
            if bind.startswith("unix:"):
                sock = socket.socket(socket.AF_UNIX, type_)
                binding = bind[5:]
                try:
                    if stat.S_ISSOCK(os.stat(binding).st_mode):
                        os.remove(binding)
                except FileNotFoundError:
                    pass
            elif bind.startswith("fd://"):
                sock = socket.socket(fileno=int(bind[5:]))
                actual_type = sock.getsockopt(socket.SOL_SOCKET, socket.SO_TYPE)
                if actual_type != type_:
                    raise SocketTypeError(type_, actual_type)
            else:
                bind = bind.replace("[", "").replace("]", "")
                try:
                    value = bind.rsplit(":", 1)
                    host, port = value[0], int(value[1])
                except (ValueError, IndexError):
                    host, port = bind, 8000
                sock = socket.socket(socket.AF_INET6 if ":" in host else socket.AF_INET, type_)

                if type_ == socket.SOCK_STREAM:
                    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)

                if self.workers > 1:
                    try:
                        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
                    except AttributeError:
                        pass
                binding = (host, port)

            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

            if bind.startswith("unix:"):
                if self.umask is not None:
                    current_umask = os.umask(self.umask)
                sock.bind(binding)
                if self.user is not None and self.group is not None:
                    os.chown(binding, self.user, self.group)
                if self.umask is not None:
                    os.umask(current_umask)
            elif bind.startswith("fd://"):
                pass
            else:
                sock.bind(binding)

            sock.setblocking(False)
            try:
                sock.set_inheritable(True)
            except AttributeError:
                pass
            sockets.append(sock)
        return sockets

    def response_headers(self, protocol: str) -> List[Tuple[bytes, bytes]]:
        headers = []
        if self.include_date_header:
            headers.append((b"date", format_date_time(time()).encode("ascii")))
        if self.include_server_header:
            headers.append((b"server", f"hypercorn-{protocol}".encode("ascii")))

        for alt_svc_header in self.alt_svc_headers:
            headers.append((b"alt-svc", alt_svc_header.encode()))
        if len(self.alt_svc_headers) == 0 and self._quic_addresses:
            from aioquic.h3.connection import H3_ALPN

            for version in H3_ALPN:
                for addr in self._quic_addresses:
                    port = addr[1]
                    headers.append((b"alt-svc", b'%s=":%d"; ma=3600' % (version.encode(), port)))

        return headers

    def set_statsd_logger_class(self, statsd_logger: Type[Logger]) -> None:
        if self.logger_class == Logger and self.statsd_host is not None:
            self.logger_class = statsd_logger

    @classmethod
    def from_mapping(
        cls: Type["Config"], mapping: Optional[Mapping[str, Any]] = None, **kwargs: Any
    ) -> "Config":
        """Create a configuration from a mapping.

        This allows either a mapping to be directly passed or as
        keyword arguments, for example,

        .. code-block:: python

            config = {'keep_alive_timeout': 10}
            Config.from_mapping(config)
            Config.from_mapping(keep_alive_timeout=10)

        Arguments:
            mapping: Optionally a mapping object.
            kwargs: Optionally a collection of keyword arguments to
                form a mapping.
        """
        mappings: Dict[str, Any] = {}
        if mapping is not None:
            mappings.update(mapping)
        mappings.update(kwargs)
        config = cls()
        for key, value in mappings.items():
            try:
                setattr(config, key, value)
            except AttributeError:
                pass

        return config

    @classmethod
    def from_pyfile(cls: Type["Config"], filename: FilePath) -> "Config":
        """Create a configuration from a Python file.

        .. code-block:: python

            Config.from_pyfile('hypercorn_config.py')

        Arguments:
            filename: The filename which gives the path to the file.
        """
        file_path = os.fspath(filename)
        spec = importlib.util.spec_from_file_location("module.name", file_path)
        module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(module)
        return cls.from_object(module)

    @classmethod
    def from_toml(cls: Type["Config"], filename: FilePath) -> "Config":
        """Load the configuration values from a TOML formatted file.

        This allows configuration to be loaded as so

        .. code-block:: python

            Config.from_toml('config.toml')

        Arguments:
            filename: The filename which gives the path to the file.
        """
        file_path = os.fspath(filename)
        with open(file_path, "rb") as file_:
            data = tomllib.load(file_)
        return cls.from_mapping(data)

    @classmethod
    def from_object(cls: Type["Config"], instance: Union[object, str]) -> "Config":
        """Create a configuration from a Python object.

        This can be used to reference modules or objects within
        modules for example,

        .. code-block:: python

            Config.from_object('module')
            Config.from_object('module.instance')
            from module import instance
            Config.from_object(instance)

        are valid.

        Arguments:
            instance: Either a str referencing a python object or the
                object itself.

        """
        if isinstance(instance, str):
            try:
                instance = importlib.import_module(instance)
            except ImportError:
                path, config = instance.rsplit(".", 1)
                module = importlib.import_module(path)
                instance = getattr(module, config)

        mapping = {
            key: getattr(instance, key)
            for key in dir(instance)
            if not isinstance(getattr(instance, key), types.ModuleType) and not key.startswith("__")
        }
        return cls.from_mapping(mapping)
